src/stub/preprocessor/forward_declarations.rs

```diff
@@ -0,0 +1,120 @@
+use super::Renderable;
+use crate::stub::{Cmd, Stub, VarType, VariableCommand};
+
+const ALPHABET: [char; 18] = [
+    'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+];
+
+/// Change the Stub structure into: [ReadDeclarations, MainContents(old_cmds)]
+/// This is relevant for Pascal.
+#[derive(Debug, Clone)]
+struct MainWrapper {
+    // Read declarations that should go on top of the main function.
+    // render declaration: int c;
+    // render read (usual): int c;\nscanf("%d", c);
+    pub forward_declarations: Vec<VariableCommand>,
+    // The main function contents.
+    pub main_content: Vec<Cmd>,
+}
+
+/// Edit stub to allow for Pascal-style forward declarations.
+///
+/// Wraps all of the commands in a stub that contains:
+/// - Forward declarations
+/// - The rest of the code
+///
+/// Traverses through the stub commands, taking all declared variables.
+/// Leaves only one MainWrapper command in the stub.
+/// Introduces two new templates:
+/// - `forward_declaration` - similar to a read_one declares one single
+///   variable, includes all the fields in a VariableCommand (not nested under
+///   `var`)
+/// - `main_wrapper` - wraps all of the code, contains `forward_declarations`
+///   (the above resource, rendered) and `main_content`
+pub fn transform(stub: &mut Stub) {
+    let mut max_nested_depth = 0;
+
+    let mut forward_declarations: Vec<VariableCommand> = stub
+        .commands
+        .iter()
+        .filter_map(|cmd| {
+            let (cmd, nested_depth) = unpack_cmd(cmd, 0);
+
+            if nested_depth > max_nested_depth {
+                max_nested_depth = nested_depth;
+            }
+
+            if let Cmd::Read(var_cmds)
+            | Cmd::LoopLine {
+                variables: var_cmds, ..
+            } = cmd
+            {
+                Some(var_cmds.into_iter())
+            } else {
+                None
+            }
```

Comment:
This would be bit shorter and (arguably?) more readable as
```suggestion
            match cmd {
                Cmd::LoopLine{ variables: var_cmds, ..} => Some(var_cmds.into_iter()),
                Cmd::Read(var_cmds) => Some(var_cmds.into_iter()),
                _ => None,
            }
```